### PR TITLE
handle errors within frames

### PR DESCRIPTION
**Pull request by jumale**

### CODE REVIEW

File: src/exporter/exporter/csv_test.go [hunk 1]
```diff
@@ -88,24 +85,23 @@ func TestNormalValues(t *testing.T) {
 		t.Run(fmt.Sprintf("Line %d", k), func(t *testing.T) {
 			select {
```
Comment (maxlaverse) [inline src/exporter/exporter/csv_test.go]: Same. No need for the `select` anymore

File: src/exporter/exporter/csv_test.go [hunk 2]
```diff
@@ -53,10 +51,9 @@ func TestRawValues(t *testing.T) {
...
```
Comment (maxlaverse) [inline src/exporter/exporter/csv_test.go]: We can drop the `select`

File: src/blackbox/reader_frame.go [hunk 3]
```diff
@@ -90,95 +90,102 @@ func NewFrameReader(dec *stream.Decoder, frameDef LogDefinition, opts *FrameRead
...
 
 	default:
+		frame.setError(frameErrorUnsupportedType(frameType))
 		//TODO: If a P frame is corrupt here, we should optionally drop the previous one (As the original implementation does)
 		f.flightLogReaderInvalidateStream()
-		f.Stats.TotalCorruptedFrames++
-		return nil, errors.Errorf("Frame type '%s' (b'%b') is not supported", string(command), command)
 	}
+
+	// if frame had any errors on reading/parsing stage - count as corrupted and return
```
Comment (maxlaverse) [inline src/blackbox/reader_frame.go]: I just had a look. You can actually drop those two `if frame.Error() != nil` and later improve the `PreComplete` method to handle that case

File: src/blackbox/reader_frame.go [hunk 4]
```diff
@@ -90,95 +90,102 @@ func NewFrameReader(dec *stream.Decoder, frameDef LogDefinition, opts *FrameRead
 }
 
 // ReadNextFrame reads the next frame
-func (f *FrameReader) ReadNextFrame() (Frame, error) {
+func (f *FrameReader) ReadNextFrame() Frame {
+	// create some base frame to make sure we always have something to return
+	// even if we could not read any frame data
+	var frame Frame = &baseFrame{}
```
Comment (maxlaverse) [inline src/blackbox/reader_frame.go]: Errors happen very rarely. We can allocate the structure only if needed

**Review by maxlaverse — APPROVED** [review]
Few minor things

File: src/blackbox/reader.go [hunk 5]
```diff
@@ -64,23 +61,27 @@ func (f *FlightLogReader) LoadFile(file io.Reader, ctx context.Context) (<-chan
...
+			frame := frameReader.ReadNextFrame()
+			// if the frame is corrupted - we should skip all bytes until the next frame
+			if frame.Error() != nil {
 				_, err = consumeToNext(decoder) // @TODO: consume skippedFrames value if needed
-				if err != nil {
-					errChan <- err
-				}
-				continue
+			}
+			if err != nil {
```
Comment (maxlaverse) [inline src/blackbox/reader.go]: What do you think of
```
if frame.Error() != nil {
  // Try to consume the remaining bytes from the broken frame
  if err := consumeToNext(decoder); err!=nil{
    frame.setError(err)
    frameChan <- frame
    return
  }
}
```
?

That way we don't need to understand what's done with the method-scoped `err` variable.

Comment (jumale) [reply src/blackbox/reader.go]: btw, question - do we plan to consume that int64 value, returned from `consumeToNext`?
https://github.com/maxlaverse/blackbox-library/blob/master/src/blackbox/reader.go#L106
it was there before, and I didn't remove it, but we actually never used it

Comment (maxlaverse) [reply src/blackbox/reader.go]: We should track the number of bytes that were dropped but where and how I'm not sure yet.

Most of statistics are generated in `FrameReader.PreComplete()`. We could move `consumeToNext()` in `FrameReader` to have them all there, and update a counter with the number of dropped bytes.

Another solution would be to generate the statistics in `FlightLogReader` instead, and move the statistics there.

Or something in the middle, move `consumeToNext()` to `FrameReader` but have it return a Frame with the remaining bytes, and the right `Start` and `Stop` offset for `FlightLogReader` to update its statistics

**Review by maxlaverse — APPROVED** [review]

**Pull request merged**